contracts/exchange/contracts/src/libs/LibExchangeRichErrorDecoder.sol

```diff
@@ -0,0 +1,443 @@
+        )
+    {
+        _assertSelectorBytes(encoded, NEGATIVE_SPREAD_ERROR_SELECTOR);
+        leftOrderHash = _readErrorParameterAsBytes32(encoded, 0);
+        rightOrderHash = _readErrorParameterAsBytes32(encoded, 1);
+    }
+
+    /// @dev Decompose an ABI-encoded TransactionError.
+    /// @param encoded ABI-encoded revert error.
+    /// @return errorCode The error code.
+    /// @return transactionHash Hash of the transaction.
+    function decodeTransactionError(bytes memory encoded)
+        public
+        pure
+        returns (
+            TransactionErrorCodes errorCode,
+            bytes32 transactionHash
+        )
+    {
+        _assertSelectorBytes(encoded, TRANSACTION_ERROR_SELECTOR);
+        errorCode = TransactionErrorCodes(_readErrorParameterAsUint256(encoded, 0));
+        transactionHash = _readErrorParameterAsBytes32(encoded, 1);
+    }
+
+    /// @dev Decompose an ABI-encoded TransactionSignatureError.
+    /// @param encoded ABI-encoded revert error.
+    /// @return transactionHash Hash of the transaction.
+    /// @return signer Signer of the transaction.
+    /// @return signature Full signature for the transaction.
+    function decodeTransactionSignatureError(bytes memory encoded)
+        public
+        pure
+        returns (
+            bytes32 transactionHash,
+            address signer,
+            bytes memory signature
+        )
+    {
+        _assertSelectorBytes(encoded, TRANSACTION_SIGNATURE_ERROR_SELECTOR);
+        transactionHash = _readErrorParameterAsBytes32(encoded, 0);
+        signer = _readErrorParameterAsAddress(encoded, 1);
+        signature = _readErrorParameterAsBytes(encoded, 2);
+    }
+
+    /// @dev Decompose an ABI-encoded TransactionExecutionError.
+    /// @param encoded ABI-encoded revert error.
+    /// @return transactionHash Hash of the transaction.
+    /// @return errorData Error thrown by exeucteTransaction().
+    function decodeTransactionExecutionError(bytes memory encoded)
+        public
+        pure
+        returns (
+            bytes32 transactionHash,
+            bytes memory errorData
+        )
+    {
+        _assertSelectorBytes(encoded, TRANSACTION_EXECUTION_ERROR_SELECTOR);
+        transactionHash = _readErrorParameterAsBytes32(encoded, 0);
+        errorData = _readErrorParameterAsBytes(encoded, 1);
+    }
+
+    /// @dev Decompose an ABI-encoded IncompleteFillError.
+    /// @param encoded ABI-encoded revert error.
+    /// @return orderHash Hash of the order being filled.
+    function decodeIncompleteFillError(bytes memory encoded)
+        public
+        pure
+        returns (
+            bytes32 orderHash
+        )
+    {
+        _assertSelectorBytes(encoded, INCOMPLETE_FILL_ERROR_SELECTOR);
+        orderHash = _readErrorParameterAsBytes32(encoded, 0);
+    }
+
+    /// @dev Revert if the leading 4 bytes of `encoded` is not `selector`.
+    function _assertSelectorBytes(bytes memory encoded, bytes4 selector)
```

Comment:
I like the `_` prefix for private members, although I don't believe it's used elsewhere in our codebase. Down to standardize this if everyone else is onboard, in which case we should add it to the internal style guidelines doc.

Reply:
I think @abandeali1 is planning on switching the whole codebase over to use the `_` prefix for internal/privates. I can switch to the current convention if this is no longer on the agenda.

Reply:
Yep, planning on doing a sweep to add the `_` to all private/internal functions at some point.